include/zephyr/bluetooth/audio/cap.h

```diff
@@ -46,6 +48,373 @@ extern "C" {
+				       int err, struct bt_conn *conn);
+
+	/**
+	 * @brief Callback for bt_cap_initiator_unicast_audio_update().
+	 *
+	 * @param unicast_group  The unicast group pointer supplied to
+	 *                       bt_cap_initiator_unicast_audio_update().
+	 * @param err            0 if success, else BT_GATT_ERR() with a
+	 *                       specific ATT (BT_ATT_ERR_*) error code.
+	 * @param conn           Pointer to the connection where the error
+	 *                       occurred. NULL if @p err is 0.
+	 */
+	void (*unicast_update_complete)(struct bt_audio_unicast_group *unicast_group,
+					int err, struct bt_conn *conn);
+
+	/**
+	 * @brief Callback for bt_cap_initiator_unicast_audio_stop().
+	 *
+	 * @param unicast_group  The unicast group pointer supplied to
+	 *                       bt_cap_initiator_unicast_audio_stop().
+	 * @param err            0 if success, else BT_GATT_ERR() with a
+	 *                       specific ATT (BT_ATT_ERR_*) error code.
+	 * @param conn           Pointer to the connection where the error
+	 *                       occurred. NULL if @p err is 0.
+	 */
+	void (*unicast_stop_complete)(struct bt_audio_unicast_group *unicast_group,
+				      int err, struct bt_conn *conn);
+#endif /* CONFIG_BT_AUDIO_UNICAST_CLIENT */
+};
+
+/**
+ * @brief Discovers audio support on a remote device.
+ *
+ * This will discover the Common Audio Service (CAS) on the remote device, to
+ * verify if the remote device supports the Common Audio Profile.
+ *
+ * @param conn Connection to a remote server.
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+int bt_cap_initiator_unicast_discover(struct bt_conn *conn);
+
+/** Type of CAP set */
+enum bt_cap_set_type {
+	/** The set is an ad-hoc set */
+	BT_CAP_SET_TYPE_AD_HOC,
+	/** The set is a CSIP Coordinated Set */
+	BT_CAP_SET_TYPE_CSIP,
+};
+
+/** Represents a Common Audio Set member that are either in a Coordinated or ad-hoc set */
+struct bt_cap_set_member {
+	union {
+		/** @brief Connection pointer if the type is BT_CAP_SET_TYPE_AD_HOC. */
+		struct bt_conn *member;
+
+		/** CSIP Coordinated Set struct used if type is BT_CAP_SET_TYPE_CSIP. */
+		struct {
+			struct bt_csis_client_set_member *member;
+		} csip;
+	};
+};
+
+struct bt_cap_stream {
+	struct bt_audio_stream bap_stream;
+	struct bt_audio_stream_ops *ops;
+};
+
+/** @brief Register Audio operations for a Common Audio Profile stream.
+ *
+ *  Register Audio operations for a stream.
+ *
+ *  @param stream Stream object.
+ *  @param ops    Stream operations structure.
+ */
+void bt_cap_stream_ops_register(struct bt_cap_stream *stream,
+				struct bt_audio_stream_ops *ops);
+
+struct bt_cap_unicast_audio_start_param {
+	/** The type of the set. */
+	enum bt_cap_set_type type;
+
+	/** The number of set members in @p members and number of streams in @p streams. */
+	size_t count;
+
+	/** Coordinated or ad-hoc set members. */
+	struct bt_cap_set_member **members;
```

Comment:
Could be moved to `struct bt_cap_stream`

Reply:
What advantage do you see from doing that? 

The `conn` pointer is already existing in the `bap_stream` of the `cap_stream`, so we would duplicate that information. 

The `struct bt_cap_stream` is currently not a struct that the application would need to set anything in (in fact, it's better if they don't), and it is something we may want to make opaque (ditto with the `bt_audio_stream`) at some point, as it is confusing when a struct contains fields that is set by both the stack and the application.

Reply:
As written:
```
	 * stream[i] will be associated with members[i] if not already
	 * initialized, else the stream will be verified against the member.
```

So the size of array of pointers are equal. Each of the member has associated stream - or each stream has a member it's addressed to. If the both structures are related why one of it cannot be a member of another one?

Reply:
This part is actually fixed by one of the next PRs: https://github.com/zephyrproject-rtos/zephyr/pull/48101/files#diff-6296d64a437af1102f200371faeed31fbc15686ff929ee09eb4afd26ce108a13R158-R190:

```c
struct bt_cap_unicast_audio_start_stream_param {
	/** Coordinated or ad-hoc set member. */
	struct bt_cap_set_member member;

	/** @brief Stream for the @p member */
	struct bt_cap_stream *stream;

	/** Endpoint reference for the @p stream */
	struct bt_audio_ep *ep;

	/**
	 * @brief Codec configuration.
	 *
	 * The @p codec.meta shall include a list of CCIDs
	 * (@ref BT_AUDIO_METADATA_TYPE_CCID_LIST) as well as a non-0
	 * stream context (@ref BT_AUDIO_METADATA_TYPE_STREAM_CONTEXT) bitfield.
	 */
	struct bt_codec *codec;

	/** Quality of Service configuration. */
	struct bt_codec_qos *qos;
};

struct bt_cap_unicast_audio_start_param {
	/** The type of the set. */
	enum bt_cap_set_type type;

	/** The number of parameters in @p stream_params */
	size_t count;

	/** Array of stream parameters */
	struct bt_cap_unicast_audio_start_stream_param *stream_params;
};
```

I can fix it in this PR as well if you wish